core/field.js

```diff
@@ -492,42 +499,172 @@ Blockly.Field.prototype.updateWidth = function() {
+      var textTool = canvas.getContext('2d');
+      Blockly.Field.cacheTextTools_[className] = textTool;
+
+      // Using CanvasRendering2DContext.measureText does not have the DOM style
+      // and layout cost that getComputedTextLength has. getComputedStyle does
+      // have that cost though. So calling initTextTools with all the classes to
+      // initialize reduces the times we'll incure the style and layout cost for
+      // getComputedStyle.
+      var computedStyle = window.getComputedStyle(clones[i]);
+
+      textTool.font = computedStyle.font;
+      if (textTool.font === Blockly.Field.CANVAS_CONTEXT_SPEC_DEFAULT_FONT) {
+        // Some browsers do not produce a computedStyle['font'] value or accept
+        // that a similarly constructed value on CanvasRendering2DContext.font
+        // member. They might accept the font reduced to weight, size and
+        // family.
+        textTool.font = (
+          computedStyle['font-weight'] + ' ' +
+          computedStyle['font-size'] + ' ' +
+          computedStyle['font-family']
+        );
+      }
+      if (textTool.font === Blockly.Field.CANVAS_CONTEXT_SPEC_DEFAULT_FONT) {
+        throw new Error('Could not initialiaze text measuring canvas.');
+      }
+    } catch (e) {
+      // Replace the canvas context with a mock object that always returns a
+      // width of 0.
+      Blockly.Field.cacheTextTools_[className] = Blockly.Field.defaultTextTool;
+    }
+  }
+
+  root.removeChild(parent);
+};
+
+/**
+ * Get a tool to measure the width of text for a given className.
+ * @param {string} className name of dom class whose font will be measured
+ * @param {Element} textElement relative element to detect font from
+ * @returns {Object} tool to measure text width
+ * @private
+ */
+Blockly.Field.getTextTool = function(className, textElement) {
+  if (Blockly.Field.cacheTextTools_) {
+    if (!Blockly.Field.cacheTextTools_[className]) {
+      Blockly.Field.initTextTools([className], textElement);
+    }
+
+    return Blockly.Field.cacheTextTools_[className];
+  }
+
+  return Blockly.Field.defaultTextTool;
+};
+
 /**
  * Gets the width of a text element, caching it in the process.
  * @param {!Element} textElement An SVG 'text' element.
  * @return {number} Width of element.
  */
 Blockly.Field.getCachedWidth = function(textElement) {
-  var key = textElement.textContent + '\n' + textElement.className.baseVal;
+  var className = textElement.className.baseVal;
+  var textContent = textElement.textContent;
   var width;
 
   // Return the cached width if it exists.
   if (Blockly.Field.cacheWidths_) {
-    width = Blockly.Field.cacheWidths_[key];
-    if (width) {
-      return width;
-    }
-  }
-
-  // Attempt to compute fetch the width of the SVG text element.
-  try {
-    if (goog.userAgent.IE || goog.userAgent.EDGE) {
-      width = textElement.getBBox().width;
+    if (!Blockly.Field.cacheWidths_[className]) {
+      Blockly.Field.cacheWidths_[className] = {};
```

Comment:
is it likely to hit this case? I thought this would have been initialized when initTextTools was called, but maybe I am misunderstanding the order in which things happen?

Reply:
Every run will execute this line a few times.

I can see the misunderstanding here. This PR introduces a second separate cache. And they both start with `cache` in their names.

This file already has a `Blockly.Field.cacheWidths_` cache. And the PR adds `Blockly.Field.cacheTextTools_`. `initTextTools` only works with `cacheTextTools_` and does not manage or interact with `cacheWidths_`. And `getCachedWidth` only works with `cacheWidths_` and does not manage or interact with `Blockly.Field.cacheTextTools_`. They are wholly independent caches. We can turn on one cache and leave the other off and they'll still function as expected.

We could move these caches out of Field and into helper classes that Field's class could make instances of. That could help distinguish the caches and functionality for text measuring. I'm not sure we need that. Everything should go through Field to get these values, so any new classes and modules from this would be ones Field depends on anyway.